### PR TITLE
REST server

### PR DESCRIPTION
**Pull request by fnevgeny**
Well, it has been my dream for a long time (predating my involvement with `pdfpc`) to have a smart remote for presentations: not just control slide moves, but also see the notes, draw, zoom, etc. In the time of pandemic it's less attractive, but hopefully this will pass...

We have already some functionality via `DBus`, but it's limited to localhost. This PR instead implements a REST-based server using `libsoup`.

I also coded a proof-of-concept Android app - get it [here](https://drive.google.com/file/d/18bAalzCi7LoF5DbjlugJaFUTF9qbk9GQ/view). In short:

* Run pdfpc with the `-V` flag. This will enable the server on port 8088 (can be changed with the `-p` switch).
*  Open the app. It should show the prefs/connection screen. Either
    - enter the hostname/ip and port manually, or
    - click `shift+Q` in pdfpc - a QR code will be shown (`escape` to get rid of it when done); press `Scan QR` in the app
* press the "V" (OK) button in the app.
* Do your presentation :)

What is currently implemented:
* Next/prev slide move (with respective buttons or swipes)
* Shows the current slide, notes, slide number
* Pointer or spotlight: press and hold the "target" button, move the device

| Screenshot_20201228-001554 | Screenshot_20201228-001726

### CODE REVIEW

**Comment by AndreasBilke** [comment]
Wow! I really like the idea and I think it's totally worth the effort. I have some (more architectural) remarks:

- what about authentication at the API? Something like a API token. `pdfpc` could generate one and announce it through the QR Code. *Otherwise I could imaging your students might be happy to discover your unprotected pdfpc instance within the network ;)*
- Do you think SSL and IPv6 support is necessary?
- Instead of just having the JSON API and build an Android app additionally (and iOS, and Sailfish OS and and and), we could ship some HTML/JS Code, which triggers the JSON API,  through the web server and every device with a browser can use that
- I would make feature optional. This would allow the compilation without `libsoup` and make `pdfpc` still usable on platforms which may not support that

One larger remark is about the structure of the program. Do you think it's a good idea to separate `pdfpc` from the REST server for a security gain? You could have `pdfpc` with a full fledged DBUS API and having `pdfpcd` for the network stuff, and `pdfpcd` communicates through DBUS with the main program.

This would allow the optional compilation without tons of `#ifdef` stuff in the code and allows the user so start `pdfpcd` with higher privileges for running on well-known ports and network devices.

Some off topic: What is your opinion on unit tests? pdfpc has some complexity which allows us to break features easily without noticing.

**Comment by fnevgeny** [comment]
>     * what about authentication at the API? Something like a API token. `pdfpc` could generate one and announce it through the QR Code. _Otherwise I could imaging your students might be happy to discover your unprotected pdfpc instance within the network ;)_

Sure, some kind of authentication is needed.

>     * Do you think SSL and IPv6 support is necessary?

I think IPv6 should work out of the box. SSL is probably not difficult to add, but there will be the standard issues with self-signed certificates (I don't think somebody will care to obtain a real cert just for pdfpc).

>     * Instead of just having the JSON API and build an Android app additionally (and iOS, and Sailfish OS and and and), we could ship some HTML/JS Code, which triggers the JSON API,  through the web server and every device with a browser can use that

Well, we or somebody else might write a real AJAX client, but it is a serious project. The simple Android demo was prepared in the Scratch-like language of the [MIT app inventor](http://ai2.appinventor.mit.edu). It's nice for fast prototyping, but the lack of VCS makes it nonpractical for anything serious. 

> 
>     * I would make feature optional. This would allow the compilation without `libsoup` and make `pdfpc` still usable on platforms which may not support that

OK
 
> One larger remark is about the structure of the program. Do you think it's a good idea to separate `pdfpc` from the REST server for a security gain? You could have `pdfpc` with a full fledged DBUS API and having `pdfpcd` for the network stuff, and `pdfpcd` communicates through DBUS with the main program.

I also thought about it, but decided not to do. First, the amount of real networking code is minimal; the rest is general logic code that needs to be added also to DBus (serialization of the response data, parsing of control parameters, etc). In addition, I believe DBus is not really good for sending large chunks of data (a rendered page at a high resolution). So pdfpc would need to output such data to a socket or named pipe, and pass to pdfpcd only a descriptor, and pdfpcd would fetch the data (and need to decide who - pdfpc or pdfpcd would clean in case of success/error, and what not). Last but not least, DBus is not OS neutral (e.g., it's pretty limited on macOS, as far as I understand, not to mention Windows). On the other hand, networking is something every dog has today.

> This would allow the optional compilation without tons of `#ifdef` stuff in the code and allows the user so start `pdfpcd` with higher privileges for running on well-known ports and network devices.

In fact, it should be possible to run pdfpc totally headless - or even compile it without GUI at all. Not only the "presenter" functionality can be implemented via REST, but also the "presentation". Or even multiple "presentation" screens - think of a class room with each student seeing the presentation on their own computer/tablet. In fact, not necessarily sitting in the same room (or even same continent...) 

> Some off topic: What is your opinion on unit tests? pdfpc has some complexity which allows us to break features easily without noticing.

It would be nice; but I have absolutely no experience with GUI unit tests; do you?

**Comment by AndreasBilke** [comment]
> I think IPv6 should work out of the box. SSL is probably not difficult to
> add, but there will be the standard issues with self-signed certificates
> (I don't think somebody will care to obtain a real cert just for pdfpc).

No, probably not for personal use. But if you have a larger network you
might have your own CA.

> Well, we or somebody else might write a real AJAX client, but it is a
> serious project. The simple Android demo was prepared in the Scratch-like
> language of the [MIT app inventor](http://ai2.appinventor.mit.edu). It's
> nice for fast prototyping, but the lack of VCS makes it nonpractical for
> anything serious. 

I think a small client shouldn't be problematic. Do you want to support all
the fancy GUI features pdfpc has (like highlighting) for the web interface
or just a "simple" forward/backward, slide preview with notes solution?

Or let's say it different: To you want a full replacement of the current
Vala/GTK based GUI?

> I also thought about it, but decided not to do. First, the amount of real
> networking code is minimal; the rest is general logic code that needs to
> be added also to DBus (serialization of the response data, parsing of
> control parameters, etc). In addition, I believe DBus is not really good
> for sending large chunks of data (a rendered page at a high resolution).
> So pdfpc would need to output such data to a socket or named pipe, and
> pass to pdfpcd only a descriptor, and pdfpcd would fetch the data (and
> need to decide who - pdfpc or pdfpcd would clean in case of
> success/error, and what not). Last but not least, DBus is not OS neutral
> (e.g., it's pretty limited on macOS, as far as I understand, not to
> mention Windows). On the other hand, networking is something every dog
> has today.

I am not familiar with inter-process communication outside of the Unix
world so I can't say something useful about OS-neutral communication. But
having network stuff in pdfpc opens a tons of new vectors for security
problems. Before, pdfpc was just on it's own and the worst thing that could
happen was malicious PDFs. But networking is a whole new topic therefore
splitting pdfpc and pdfpcd might make me sleep well :)

> In fact, it should be possible to run pdfpc totally headless - or even
> compile it without GUI at all. Not only the "presenter" functionality can
> be implemented via REST, but also the "presentation". Or even multiple
> "presentation" screens - think of a class room with each student seeing
> the presentation on their own computer/tablet. In fact, not necessarily
> sitting in the same room (or even same continent...) 

Maybe I think in other use cases for the REST server than you, but this sounds
more like reinventing the wheel. Screen sharing is nothing new. I was
thinking about a serious remote control possibility for pdfpc where you are
not bound to your laptop but can have a tablet during your talk.

> It would be nice; but I have absolutely no experience with GUI unit tests; do you?

I am familiar with non-GUI unit tests.

**Comment by fnevgeny** [comment]
> No, probably not for personal use. But if you have a larger network you might have your own CA.

There are certainly various use cases. The one I primarily had in mind is giving talks at conferences (not huge ones, though). You come with your laptop and smartphone, both are connected to a guest WiFi network. In which case the IP and FQDN of the laptop are unpredictable, so no certificate would be valid. It's possible to use the laptop as an AP (or build a VPN atop of the guest network), but then you control the access at a more basic level, no need for extra layers of protection.

> I think a small client shouldn't be problematic. Do you want to support all the fancy GUI features pdfpc has (like highlighting) for the web interface or just a "simple" forward/backward, slide preview with notes solution?

Personally, I want the whole functionality I currently use during the presentations - which is almost all minus note editing. Yes, including highlighting, zooming-in, and even simple drawing.

> I am not familiar with inter-process communication outside of the Unix world so I can't say something useful about OS-neutral communication. But having network stuff in pdfpc opens a tons of new vectors for security problems. Before, pdfpc was just on it's own and the worst thing that could happen was malicious PDFs. But networking is a whole new topic therefore splitting pdfpc and pdfpcd might make me sleep well :)

Well, everything comes at its price :). However, given the networking functionality is off by default, i.e., requires a knowledgeable decision on the side of user to enable it, you probably may continue sleeping well most of the time... Of course, as discussed, a kind of authorization token (and SSL if desired) will be there. I also consider adding a "lock-in" mechanism: right after scanning the QR code the client sends a one-time confirmation message and since that moment pdfpc serves only that single address. 

BTW, what happened to your previous less-than-warm feelings about DBus? :)

> Maybe I think in other use cases for the REST server than you, but this sounds more like reinventing the wheel. Screen sharing is nothing new.

It isn't a screen sharing. First, screen sharing requires a low-latency and high-bandwidth connection. While in our case, it's like web browsing - you fetch a new image once in a while (typically 1-2 minutes), and that's all. But there is more to it in functionality. With screen sharing, you (student, listener) is strictly limited to what the presenter shows. With a REST client, you may want to go a few slides back to recall something (no need to interrupt the lecturer "can you please go to the previous slide" - when you are allowed to, at all), zoom in a part according to *your* visual acuity, or save a slide in high resolution (snapshotting off a video stream is quite awful). I can think of more interesting possibilities, e.g., providing a valuable feedback to the lecturer at the end or in real time (e.g., a statistics of how many students switched back to an earlier slide - a hint it might need a longer explanation, etc).

> I was thinking about a serious remote control possibility for pdfpc where you are not bound to your laptop but can have a tablet during your talk.

Yes, this was (and still is) my primary focus - even more specifically, not a tablet but smartphone (to hold it in palm), but why not think about other ways towards world domination? :)

> I am familiar with non-GUI unit tests.

Same here.

Happy New Year!

**Comment by AndreasBilke** [comment]
> Personally, I want the whole functionality I currently use during the
> presentations - which is almost all minus note editing. Yes, including
> highlighting, zooming-in, and even simple drawing.

Wouldn't that make the current GTK implementation obsolete or results in
double effort to make it feature tie?

> Well, everything comes at its price :). However, given the networking
> functionality is off by default, i.e., requires a knowledgeable decision
> on the side of user to enable it, you probably may continue sleeping well
> most of the time... Of course, as discussed, a kind of authorization
> token (and SSL if desired) will be there. I also consider adding a
> "lock-in" mechanism: right after scanning the QR code the client sends a
> one-time confirmation message and since that moment pdfpc serves only
> that single address. 

Sounds better :) But still leaves some #ifdef in the code to keep the
feature optional.

> BTW, what happened to your previous less-than-warm feelings about DBus?
> :)

It's still cold but probably the best choice for IPC to have pdfpc and
pdfpcd seperated :)

Happy new year!

**Comment by fnevgeny** [comment]
> Wouldn't that make the current GTK implementation obsolete or results in double effort to make it feature tie?

I don't think it will make the Gtk GUI obsolete. The Web GUI, and in particular the handheld device variant is not for everyone/every use case. There will be some duplication of efforts, but it will be on the client side (which should be a separate project). The REST glue itself in pdfpc is a minor part. On the other hand, making both the native and Web GUI's work will likely enforce a cleaner MVC code base structuring.

> Sounds better :) But still leaves some #ifdef in the code to keep the feature optional.

#ifdef's are not for extra security, but only because you suspect libsoup might not be available on every build target. When compiled but not activated, the server code itself should not be a threat, I believe...

**Comment by fnevgeny** [comment]
By and large, the security measures are now implemented (password, TLS, lock-in). If password is not set, a random one will be generated. For TLS to work, cert.pem and key.pem files must be present in the user's pdfpc config directory.

I don't think splitting the REST functionality in a standalone pdfpcd would significantly improve security. Both pdfpc and pdfpcd would run under the same userid, so a bug in libsoup or in our handling of REST input/output would bring the same consequences. Such a splitting makes sense to achieve privilege separation, but it isn't our case...

**Comment by AndreasBilke** [comment]
For the easy case you are right, but you could not run the REST server
under a privileged port without being root (which probably makes running
X-Server very difficult), also you can not put your certificate under
/etc/ssl/... without root.

But I must admit this is probably not very comming for that use case here.

Are you okay with the code? Do we want to give it a try in the wild?

**Comment by fnevgeny** [comment]
> For the easy case you are right, but you could not run the REST server under a privileged port without being root (which probably makes running X-Server very difficult), also you can not put your certificate under /etc/ssl/... without root. 

Oh no, it should never run under a privileged account. Just like the (user-level) `dbus-daemon`. Essentially the same functionality, but a networked protocol.

> Are you okay with the code? Do we want to give it a try in the wild?

Not yet, there are some issues with CORS...

**Comment by fnevgeny** [comment]
Well, I think it's largely ready for merging... Some aspects are not implemented yet, e.g., timer and pointer events, but these need more substantial changes in the core of pdfpc itself, so better to be postponed.

File: man/pdfpc.in [hunk 1]
```diff
@@ -70,6 +70,9 @@ Allow for testing pdfpc without proper installation. Icons, configuration files,
 etc. will be loaded from the source path locations. Mostly intended for
 developers.
 .TP
+.BI "\-p, \-\-rest\-port"
+Port the REST server binds to. (Default: 8088)
```
Comment (AndreasBilke) [inline man/pdfpc.in:74]: Is it on purpose to use "radan http" port? https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml?search=8088

File: src/classes/rest_server.vala [hunk 2]
```diff
@@ -0,0 +1,594 @@
...
+                string response = generator.to_data(null);
+
+                msg.set_response("application/json", Soup.MemoryUse.COPY,
+                    response.data);
+            }
+        }
+
+        public RestServer(Metadata.Pdf metadata, int port_num) {
+            TlsCertificate cert = null;
+            if (Options.rest_https) {
```
Comment (AndreasBilke) [inline src/classes/rest_server.vala:456]: If I understand it correctly, you would use the same port for HTTP and HTTPS traffic. Is this intentional?

**Review by AndreasBilke — APPROVED** [review]
I'm fine with it now (the comments are just hints).

If you want you can merge it.

File: man/pdfpc.in (hunk 1 above)
Comment (fnevgeny) [reply man/pdfpc.in:74]: No, not really. I've never heard about it, frankly. Could be any other, as long as it is an unprivileged port and not typically used on an average PC...

File: src/classes/rest_server.vala (hunk 2 above)
Comment (fnevgeny) [reply src/classes/rest_server.vala:456]: Yes. It is anyway not possible to run HTTP and HTTPS simultaneously (can be implemented, but I see absolutely no need), so why make it more complicated?

**Comment by AndreasBilke** [comment]
> Yes. It is anyway not possible to run HTTP and HTTPS simultaneously (can be implemented, but I see absolutely no need), so why make it more complicated?

Just because it is uncommon in other applications (80/443, 8080/8081).
There is no technical need for it.

**Comment by fnevgeny** [comment]
> Just because it is uncommon in other applications (80/443, 8080/8081). There is no technical need for it.

There are of course many reasons for HTTP(S) to run on different default ports, including legacy considerations. I'd rather not bother at this stage with choosing potentially saner defaults for pdfpc, let alone applying for registering an unassigned port(s) with IANA...

**Review by AndreasBilke — APPROVED** [review]

**Pull request merged**